src/cli/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR BSD-3-Clause
+
+#![cfg(target_arch = "x86_64")]
+
+//! CLI for the Reference VMM.
+
+#![deny(missing_docs)]
+
+extern crate clap;
+extern crate vmm;
+
+use std::convert::TryFrom;
+use std::result;
+
+use clap::{App, Arg};
+use vmm::{KernelConfig, MemoryConfig, VMMConfig, VcpuConfig};
+
+/// Command line parser.
+pub struct CLI;
+
+impl CLI {
+    /// Parses the command line options into VMM configurations.
+    ///
+    /// # Arguments
+    ///
+    /// * `app_name` - name of the running program.
+    pub fn launch(cmdline_args: Vec<&str>) -> result::Result<VMMConfig, String> {
+        let matches = App::new(cmdline_args[0].to_string())
+            .arg(
+                Arg::with_name("memory")
+                    .long("memory")
+                    .required(true)
+                    .takes_value(true)
+                    .validator(Self::validate_memory_config),
+            )
+            .arg(
+                Arg::with_name("vcpus")
+                    .long("vcpus")
+                    .required(true)
+                    .takes_value(true)
+                    .validator(Self::validate_vcpu_config),
+            )
+            .arg(
+                Arg::with_name("kernel")
+                    .long("kernel")
+                    .required(true)
+                    .takes_value(true)
+                    .validator(Self::validate_kernel_config),
+            )
+            .get_matches_from_safe(cmdline_args)
+            .map_err(|_| "Failed to parse command line arguments".to_string())?;
```

Comment:
I haven't figured out a way to do either. To me it seemed like the easiest way to validate those strings was to try and cast them into `*Config` objects, and consider them valid if the cast succeeds. As this CLI is meant to be for demonstrative purposes only, and throw-away, I didn't think too much of anything fancier :smile: The cleanest way IMO would be to replace `clap` like you mentioned in another comment, in the meantime I think we can compromise on the CLI at least until we commit to something as production-worthy as the VMM - WDYT?

Reply:
You're absolutely right that the user experience is pretty crappy and needs more work, and thanks for the reminder of why we're doing this in the first place :smile: 
I'm working on nicer errors.
> call `get_matches_from()` instead of `get_matches_from_safe()`

The `_safe` one returns a testable `Result`. The other one `exit`s the process, so there's no way to unit test the CLI interaction. How about a compromise: keep the `_safe()` one and the tests, but print the usage in case of errors?